server/pkcs11.go

```
@@ -0,0 +1,158 @@
+	var qAttrs []string
+	var parts  []string
+	var value string
+
+	uri = strings.Split(uri, "pkcs11:")[1]
+	parts = strings.Split(uri, "?")
+	pAttrs = strings.Split(parts[0], ";")
+	if 1 < len(parts) {
+		qAttrs = strings.Split(parts[1], "&")
+	}
+
+	for _, attr := range pAttrs {
+		parts = strings.Split(attr, "=")
+		parts[0] = strings.Trim(parts[0], " \n\t\r")
+		if 1 < len(parts) {
+			parts[1] = strings.Trim(parts[1], " \n\t\r")
+			value, _ = url.QueryUnescape(parts[1])
+		}
+		switch parts[0] {
+		case "token":
+			pk11uri.Token = value
+		case "manufacturer":
+			pk11uri.Manuf = value
+		case "serial":
+			pk11uri.Serial = value
+		case "model":
+			pk11uri.Model = value
+		case "library-manufacturer":
+			pk11uri.LibManuf = value
+		case "library-description":
+			pk11uri.LibDesc = value
+		case "library-version":
+			pk11uri.LibVer = value // TODO 1*DIGIT [ "." 1*DIGIT ]
+		case "object":
+			pk11uri.Object = []byte(value)
+		case "type":
+			pk11uri.Type = value // TODO public, private, cert, secret-key, data
+		case "id":
+			pk11uri.Id = []byte(value)
+		case "slot-manufacturer":
+			pk11uri.SlotManuf = value
+		case "slot-description":
+			pk11uri.SlotDesc = value
+		case "slot-id":
+			id, _ := strconv.ParseUint(value, 10, 32) // TODO what is the bit size
+			pk11uri.SlotId = uint(id)
+		default:
+			log.Warning("Unrecognized PKCS#11 URI Component: " + parts[0])
+		}
+	}
+
+	for _, attr := range qAttrs {
+		// TODO: do we need to trim attr here?
+		parts = strings.Split(attr, "=")
+		parts[0] = strings.Trim(parts[0], " \n\t\r")
+		if 1 < len(parts) {
+			parts[1] = strings.Trim(parts[1], " \n\t\r")
+			value, _ = url.QueryUnescape(parts[1])
+		}
+		switch parts[0] {
+		case "pin-source":
+			pk11uri.PinSource = value
+		case "pin-value":
+			pk11uri.PinValue = value
+		case "module-name":
+			pk11uri.ModuleName = value
+		case "module-path":
+			pk11uri.ModulePath = value
+		default:
+			log.Warning("Unrecognized PKCS#11 URI Query: " + parts[0])
```

Same comment - return error instead of logging.